File: go32/adt2vesa/VESA_InitStepFade.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

void VESA_InitStepFade (const tPaletteArray StartPalette,
                        tPaletteArray EndPalette, float Speed) {
  float EndRealPal[256*3]; //0
  uint_least16_t i;

  for (i = 0; i < 256*3; i++) {
      StepRealPal[i] = (float) StartPalette[i];
      StepWorkPalette[i] = StartPalette[i];
      EndRealPal[i] = (float) EndPalette[i];
      StepDelta[i] = (EndRealPal[i] - StepRealPal[i]) / Speed;
  }

  VESA_SetPalette (StartPalette);
  FadeSteps = truncf (Speed);
}

File: go32/adt2vesa/pas/VESA_InitStepFade.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure VESA_InitStepFade(var StartPalette,EndPalette; Speed: Single);

var
  EndRealPal: array[0..767] of Single;
  idx: Word;

begin
  For idx := 0 to 767 do
    begin
      StepRealPal[idx] := tPaletteArray(StartPalette)[idx];
      StepWorkPalette[idx] := tPaletteArray(StartPalette)[idx];
      EndRealPal[idx] := tPaletteArray(EndPalette)[idx];
      StepDelta[idx] := (EndRealPal[idx]-StepRealPal[idx])/Speed;
    end;
  VESA_SetPalette(StartPalette);
  FadeSteps := TRUNC(Speed);
end;
